OpenRA.Mods.Common/Traits/World/ColorPickerManager.cs

```diff
@@ -55,74 +54,61 @@ public class ColorPickerManagerInfo : TraitInfo<ColorPickerManager>, IColorPicke
+				// and too large for bright colors. By using gamma we shift the metric to be more uniform.
+				// This hack doesn't fully fix bright colors, i.e. it still allows for very similar pinks
+				// and reports colors with slightly different saturations as significantly different.
+				// TODO: Replace with a model which has image hue remapping in mind.
+				var rmean = (color.R + candidate.R) / 2;
+				var rdelta = color.R - candidate.R;
+				var gdelta = color.G - candidate.G;
+				var bdelta = color.B - candidate.B;
+
+				if (Math.Sqrt(((512 + rmean) * rdelta * rdelta >> 8) + 4 * gdelta * gdelta + ((767 - rmean) * bdelta * bdelta >> 8)) < SimilarityThreshold)
+					return true;
 			}
 
-			return closestDistance < SimilarityThreshold;
+			return false;
 		}
 
 		Color MakeValid(float hue, float sat, float val, MersenneTwister random, IEnumerable<Color> terrainColors, IEnumerable<Color> playerColors, Action<string> onError)
-		{
-			var terrainLinear = terrainColors.Select(c => c.ToLinear()).ToList();
-			var playerLinear = playerColors.Select(c => c.ToLinear()).ToList();
-
-			return MakeValid(hue, sat, val, random, terrainLinear, playerLinear, onError);
-		}
-
-		Color MakeValid(float hue, float sat, float val, MersenneTwister random, List<(float R, float G, float B)> terrainLinear, List<(float R, float G, float B)> playerLinear, Action<string> onError)
 		{
 			// Clamp saturation without triggering a warning
 			// This can only happen due to rounding errors (common) or modified clients (rare)
 			sat = sat.Clamp(HsvSaturationRange[0], HsvSaturationRange[1]);
 			val = val.Clamp(HsvValueRange[0], HsvValueRange[1]);
 
-			// Limit to 100 attempts, which is enough to move all the way around the hue range
-			string errorMessage = null;
-			var stepSign = 0;
-			for (var i = 0; i < 101; i++)
+			string errorMessage;
+			var color = Color.FromAhsv(hue, sat, val);
+			if (IsInvalidColor(color, terrainColors))
+				errorMessage = PlayerColorTerrain;
+			else if (IsInvalidColor(color, playerColors))
+				errorMessage = PlayerColorPlayer;
+			else
+				return color;
+
+			// Move by expanding from the selected color in both directions by a limited amount circling
+			// around the hue a bunch of times. This method usually returns a color similar the selected
+			// and controls the randomness.
+			// Exit after 400 iterations to avoid infinite loops.
+			for (var i = 2; i < 402; i++)
 			{
-				var linear = Color.FromAhsv(hue, sat, val).ToLinear();
-				if (TryGetBlockingColor(linear, terrainLinear, out var blocker))
-					errorMessage = PlayerColorTerrain;
-				else if (TryGetBlockingColor(linear, playerLinear, out blocker))
-					errorMessage = PlayerColorPlayer;
-				else
-				{
-					if (errorMessage != null)
-						onError?.Invoke(errorMessage);
+				color = Color.FromAhsv(
+					hue + (i % 2 == 0 ? -1 : 1) * (i / 2) * 0.25f * (0.2f + random.NextFloat()),
```

Comment:
This causes a huge jump in hue, producing colours that are quite different (e.g. violet from pink) rather than just sufficiently different. Doesn't feel right, IMO.

Reply:
From my testing in majority of cases it returns a similar hue, and the math works out to that way. The more it misses the farther is strays from the original hue, until it is fully rotated in which case it can be whatever

Reply:
I think it's important to view it with the perspective of, our last 2 implementation of color correction were quite bad at selecting similar colors. And also with the perspective that we want the correction to provide a bit wilder guesses than the validator would deem sufficient as we know that the validator is not perfect